### PR TITLE
LPS-202418 Generic error when set read only in the relationship field after bind

### PR DESCRIPTION
**Pull request by larissacribeiro**
**Related Issue**: [**LPS-202418**](https://liferay.atlassian.net/browse/LPS-202418)

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 1]
```diff
@@ -185,6 +185,21 @@ public ObjectField addOrUpdateCustomObjectField(
 				objectFieldSettings);
 		}
 
+		if (Objects.equals(businessType, "Relationship")) {
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:188]: @larissacribeiro please use the constant instead: `ObjectFieldConstants.BUSINESS_TYPE_RELATIONSHIP`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 2]
```diff
@@ -185,6 +185,21 @@ public ObjectField addOrUpdateCustomObjectField(
...
+			ObjectDefinition objectDefinition =
+				_objectDefinitionPersistence.findByPrimaryKey(
+					objectDefinitionId);
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:191]: @larissacribeiro we have this same logic [here](https://github.com/liferay-objects/liferay-portal/blob/0e2895f791b4654956813ecd9e25f6621aede39e/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java#L170-L173).

Please move this logic to be outside the `if` statement to reuse code.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 3]
```diff
@@ -185,6 +185,21 @@ public ObjectField addOrUpdateCustomObjectField(
...
+
+			if (objectDefinition.isRootDescendantNode() &&
+				(Objects.equals(readOnly, "true") ||
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:194]: @larissacribeiro please use the constant instead: `ObjectFieldConstants.READ_ONLY_TRUE`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 4]
```diff
@@ -185,6 +185,21 @@ public ObjectField addOrUpdateCustomObjectField(
...
+				 Objects.equals(readOnly, "conditional"))) {
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:195]: @larissacribeiro please use the constant instead: `ObjectFieldConstants.READ_ONLY_CONDITIONAL`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 5]
```diff
@@ -185,6 +185,21 @@ public ObjectField addOrUpdateCustomObjectField(
...
+
+				throw new ObjectFieldReadOnlyException(
+					"The readOnly configuration is set by a root object and " +
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:198]: @larissacribeiro change to `root object definition`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java (hunk 4 above)
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:195]: @larissacribeiro the ticket describes the expected behavior as: 

> we must validate that the user cannot update this property.

So I think it makes more sense to do:  
```
if (objectDefinition.isRootDescendantNode() && 
    !Objects.equals(existingObjectField.getReadOnly(), readOnly))
```

Can we do this?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by larissacribeiro** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java (hunk 4 above)
Comment (larissacribeiro) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java]: Yes, you're right! thanks for the review @carolmariaabb, just updated it with what you requested.

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by carolmariaabb** [comment]
@larissacribeiro related Failure:

1. [com.liferay.object.service.test.ObjectFieldLocalServiceTest](https://testray.liferay.com/home/-/testray/case_results/2475888778/history?tab=result&hideFilterPin=true)

**Comment by larissacribeiro** [comment]
ci:test:sf

**Comment by larissacribeiro** [comment]
Hi @carolmariaabb ! The test was failing due to moving the ObjectDefinition logic outside of the if statement as we had done, so I've taken the new validation into the private `updateObjectField` method in order not to mess with the validations in `addOrUpdateCustomObjectField`. Is it okay?

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by carolmariaabb** [comment]
Possible related failure:
1.[LocalFile.SystemObjectAccount#CannotDeleteEntryWhenDeletionTypeIsPrevent](https://testray.liferay.com/home/-/testray/case_results/2476935430)

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:forward:force

**Pull request closed without merging**